br/pkg/backup/prepare_snap/stream.go

```diff
@@ -0,0 +1,209 @@
+	storeID       uint64
+	cli           PrepareClient
+	leaseDuration time.Duration
+
+	output       chan<- event
+	serverStream <-chan utils.Result[*brpb.PrepareSnapshotBackupResponse]
+
+	clientLoopHandle *errgroup.Group
+	stopBgTasks      context.CancelFunc
+}
+
+// InitConn initializes the connection to the stream (i.e. "active" the stream).
+//
+// Before calling this, make sure you have filled the store ID and output channel.
+//
+// Once this has been called, them **should not be changed** any more.
+func (p *prepareStream) InitConn(ctx context.Context, cli PrepareClient) error {
+	p.cli = cli
+	p.clientLoopHandle, ctx = errgroup.WithContext(ctx)
+	ctx, p.stopBgTasks = context.WithCancel(ctx)
+	return p.GoLeaseLoop(ctx, p.leaseDuration)
+}
+
+func (p *prepareStream) Finalize(ctx context.Context) error {
+	log.Info("shutting down", zap.Uint64("store", p.storeID))
+	return p.stopClientLoop(ctx)
+}
+
+func (p *prepareStream) GoLeaseLoop(ctx context.Context, dur time.Duration) error {
+	err := p.cli.Send(&brpb.PrepareSnapshotBackupRequest{
+		Ty:             brpb.PrepareSnapshotBackupRequestType_UpdateLease,
+		LeaseInSeconds: uint64(dur.Seconds()),
+	})
+	if err != nil {
+		return errors.Annotate(err, "failed to initialize the lease")
+	}
+	msg, err := p.cli.Recv()
+	if err != nil {
+		return errors.Annotate(err, "failed to recv the initialize lease result")
+	}
+	if msg.Ty != brpb.PrepareSnapshotBackupEventType_UpdateLeaseResult {
+		return errors.Errorf("unexpected type of response during creating lease loop: it is %s", msg.Ty)
+	}
+	p.serverStream = utils.AsyncStreamBy(p.cli.Recv)
+	p.clientLoopHandle.Go(func() error { return p.clientLoop(ctx, dur) })
+	return nil
+}
+
+func (p *prepareStream) onResponse(ctx context.Context, res utils.Result[*brpb.PrepareSnapshotBackupResponse]) error {
+	if err := res.Err; err != nil {
+		return err
+	}
+	resp := res.Item
+	logutil.CL(ctx).Debug("received response", zap.Stringer("resp", resp))
+	evt, ok := p.convertToEvent(resp)
+	if ok {
+		logutil.CL(ctx).Debug("generating internal event", zap.Stringer("event", evt))
+		p.output <- evt
+	}
+	return nil
+}
+
+func (p *prepareStream) stopClientLoop(ctx context.Context) error {
+	p.stopBgTasks()
+	if err := p.clientLoopHandle.Wait(); err != nil {
+		return err
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case res, ok := <-p.serverStream:
+			err := p.onResponse(ctx, res)
+			if err == io.EOF || !ok {
+				logutil.CL(ctx).Info("close loop done.", zap.Uint64("store", p.storeID), zap.Bool("is-chan-closed", !ok))
+				return nil
+			}
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (p *prepareStream) clientLoop(ctx context.Context, dur time.Duration) error {
+	ticker := time.NewTicker(dur / 4)
+	lastSuccess := time.Unix(0, 0)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			logutil.CL(ctx).Info("client loop exits.", zap.Uint64("store", p.storeID))
+			return p.cli.Send(&brpb.PrepareSnapshotBackupRequest{
```

Comment:
How about send Finish request in `stopClientLoop` since we have to `Finalize` when exits?